src/main/java/Ex16.java

```diff
@@ -0,0 +1,21 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class Ex16 {
+    public int fibonacci(int n) {
+        if (n == 0 || n == 1) {
+            return n;
+        } else {
+            return fibonacci(n - 1) + fibonacci(n - 2);
+        }
+    }
+
+    public int demFibo(int n) {
```

Comment:
Mình thường thấy người ta dùng bài toán về số fibonacci để minh họa cho việc gọi đệ quy. Tuy nhiên điều đó không có nghĩa là thuật toán đệ quy nên được sử dụng cho số fibonacci. 

Tưởng tượng cần tìm số fibonacci thứ 20, mình cần thực hiện 20 phép cộng là xong nếu mình cứ bắt đầu a=b=1 và sau đó a=a+b, b=a+b nên 1, 1 --> 2 -> 3 -> 5....

Tuy nhiên nếu mình dùng đệ quy, chuyện gì xảy ra.

Để tính fibo(20) mình đệ quy cho 19 và 18
Để tính fibo(19) mình đệ quy tính 18 và 17

Như vậy sơ sơ mình cần tính 2 lần fibo(18), 3 lần fibo(17), 5 lần fibo(16) nó còn và đến khi ra kết quả fibo(1) thì cũng là hàng tỉ lần

![fibo](https://holycoders.com/content/images/wordpress/2020/04/fibonacci-sequence-recursion-tree.png)

Do đó ở đây, đừng dùng đệ quy nữa, suy nghĩ cách khác.

Ngoài ra, ở đây cái ArrayList chỉ dùng để đếm, do đó có thể bỏ đi và thay bằng biến count thì sẽ có tác dụng tương tự